sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -414,8 +414,42 @@ class Analyzer(
     }
 
     def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
-      case i @ InsertIntoTable(u: UnresolvedRelation, _, _, _, _) =>
-        i.copy(table = EliminateSubqueryAliases(getTable(u)))
+      case i @ InsertIntoTable(u: UnresolvedRelation, parts, child, _, _) if child.resolved =>
+        val table = getTable(u)
+        // adding the table's partitions or validate the query's partition info
+        table match {
+          case relation: PartitionedRelation if relation.partitionColumns.nonEmpty =>
+            val tablePartitionNames = relation.partitionColumns.map(_.name)
+            if (parts.keys.nonEmpty) {
+              // the query's partitioning must match the table's partitioning
+              // this is set for queries like: insert into ... partition (one = "a", two = <expr>)
+              if (tablePartitionNames.size != parts.keySet.size) {
```

Comment:
I checked the `PreWriteCheck` rule, it does check if the user specified partition columns match the real ones. So I think we don't need to check here but just add a comment to say that the `PreWriteCheck` will do this check.